source/ProcessManager.Orchestrations/InternalProcesses/MigrateCalculationsFromWholesale/V1/Activities/MigrateCalculationActivity_MigrateCalculationsFromWholesale_V1.cs

```
@@ -38,10 +53,129 @@ public static Guid GetMigratedWholesaleCalculationIdCustomStateGuid(Orchestratio
+            PeriodEndDate: ToDateTimeOffset(wholesaleCalculation.PeriodEnd),
+            IsInternalCalculation: wholesaleCalculation.IsInternalCalculation);
+        orchestrationInstance.ParameterValue.SetFromInstance(calculationInput);
+
+        // Orchestration Instance => Running
+        orchestrationInstance.Lifecycle.TransitionToRunning(new MagicClock(wholesaleCalculation.ExecutionTimeStart));
+
+        // Step: Calculation
+        orchestrationInstance.TransitionStepToRunning(1, new MagicClock(wholesaleCalculation.ExecutionTimeStart));
+        orchestrationInstance.TransitionStepToTerminated(1, OrchestrationStepTerminationState.Succeeded, new MagicClock(wholesaleCalculation.ExecutionTimeEnd));
+
+        // Step: Enqueue messages
+        var stepsSkippedBySequence = orchestrationInstance.Steps
+            .Where(step => step.IsSkipped())
+            .Select(step => step.Sequence)
+            .ToList();
+        if (!stepsSkippedBySequence.Contains(2))
+        {
+            orchestrationInstance.TransitionStepToRunning(2, new MagicClock(wholesaleCalculation.ActorMessagesEnqueuingTimeStart));
+            orchestrationInstance.TransitionStepToTerminated(2, OrchestrationStepTerminationState.Succeeded, new MagicClock(wholesaleCalculation.ActorMessagesEnqueuedTimeEnd));
+        }
+
+        // Orchestration Instance => Terminated
+        orchestrationInstance.Lifecycle.TransitionToSucceeded(new MagicClock(wholesaleCalculation.CompletedTime));
+
+        return orchestrationInstance;
+    }
+
+    /// <summary>
+    /// An implementation of IClock that allows us to manipulate time and set it to a decired value.
+    /// This is used in state transitions to be able to make the transition happen at a certain time.
+    /// We do not allow nulls, and they should not appear within the given dataset for migration
+    /// </summary>
+    /// <param name="timeOrNull"></param>
+    private class MagicClock(Instant? timeOrNull) : IClock
```

Should we throw an exception if we do not allow nulls?

---

Since we only migrate "succeeded/completed" calculations, there will not be null's. The reason I have to allow null is because the "source" (Calculation) allows nulls, but we know they will always have a value.